text/3735-field-projections.md

```diff
@@ -0,0 +1,1745 @@
+
+#### Pin Projections
+
+For this section, you should understand what [pin projections] are. If not, then you can just skip
+this section.
+
+Structurally pinned fields are marked with `#[pin]` using the derive macro `PinProject`. For example
+consider a future that alternatingly polls two futures:
+
+```rust
+#[derive(PinProject)]
+struct FairRaceFuture<F1, F2> {
+    #[pin]
+    fut1: F1,
+    #[pin]
+    fut2: F2,
+    fair: bool,
+}
+```
+
+Now, it's possible to project a `fut: Pin<&mut FairRaceFuture>`:
+- `fut->fut1: Pin<&mut F1>`
+- `fut->fut2: Pin<&mut F2>`
+- `fut->fair: &mut bool`
+
+Using these, one can concisely implement `Future` for `FairRaceFuture` without any `unsafe` code:
+
+```rust
+impl<F1: Future, F2: Future<Output = F1::Output>> Future for FairRaceFuture<F1, F2> {
+    type Output = F1::Output;
+
+    fn poll(self: Pin<&mut Self>, ctx: &mut Context) -> Poll<Self::Output> {
+        let fair: &mut bool = self->fair;
+        *fair = !*fair;
+        if *fair {
+            // self->fut1: Pin<&mut F1>
+            match self->fut1.poll(ctx) {
+                Poll::Pending => self->fut2.poll(ctx),
+                Poll::Ready(res) => Poll::Ready(res),
+            }
+        } else {
+            // self->fut2: Pin<&mut F2>
+            match self->fut2.poll(ctx) {
+                Poll::Pending => self->fut1.poll(ctx),
+                Poll::Ready(res) => Poll::Ready(res),
+            }
+        }
+    }
+}
+```
+
+### Implementing Custom Field Projections
+
+There are two different ways to implement projections for a custom type:
+- implementing the `Projectable` and `Project<F>` traits, or
+- annotating it with `#[projecting]`.
+
+They are used for pointers and container types respectively.
+
+#### Pointer-Like
+
+Pointer-like types can be projected using `Projectable` and `Project<F>`. For example, if we create
+a custom reference type that simultaneously points at two instances of the same type.
+
+```rust
+pub struct DoubleRef<'a, T> {
+    first: &'a T,
+    second: &'a T,
+}
+
+impl<'a, T> DoubleRef<'a, T> {
+    pub fn new(first: &'a T, second: &'a T) -> Self {
+        Self { first, second }
+    }
+
+    pub fn read(&self) -> (T, T)
+    where
+        T: Copy
+    {
+        (*self.first, *self.second)
+    }
+}
+```
+
+We can now allow its users to use field projection by implementing the above mentioned traits:
+
+```rust
+impl<'a, T> Projectable for DoubleRef<'a, T> {
+    type Inner = T;
+}
+```
+
+The `Projectable` trait is what tells the compiler which types' fields to consider for projecting.
+When you write `a->b`, then `a` has to implement `Projectable` in order for the compiler to know
+which type to look up the field `b` for.
+
+The actual projection operation is governed by `Project<F>`:
+
+```rust
+impl<'a, T, F> Project<F> for DoubleRef<'a, T>
+where
+    // Since we want to project both of our references, we want the field to be aligned.
+    F: UnalignedField<Base = T> + Field,
```

Comment:
wait I thought using `Field<Base = T>` works?

```suggestion
    F: Field<Base = T>,
```

Reply:
yes it does! I thought it didn't and I didn't check :)

Reply:
yeah that's the whole reason why I proposed the names `UnalignedField` and `Field` :)

if the supertrait has to be explicitly spelled out I would have called them `Field` and `AlignedField`